docs/source/notes/randomness.rst

```diff
@@ -2,70 +2,73 @@ Reproducibility
-that all other libraries your code relies on and which use random numbers also
-use a fixed seed.
+PyTorch uses a pseudorandom number generator, which you will need to seed
+manually to make it act deterministically. Furthermore, you can configure
+PyTorch to avoid using nondeterministic algorithms for a subset of operations.
+Finally, if your code relies on other libraries that use random numbers, make
+sure to give them a fixed seed.
 
-PyTorch
-.......
+Random number generation
+........................
 You can use :meth:`torch.manual_seed()` to seed the RNG for all devices (both
 CPU and CUDA)::
 
     import torch
     torch.manual_seed(0)
 
+Deterministic algorithms
+........................
+Some PyTorch operations use nondeterministic algorithms by default. A subset
+of them can be configured to use a deterministic implementation or throw an
+error if a deterministic implementation is not available. Please check the
+documentation for :meth:`torch.set_deterministic()` for a full list of affected
+operations. If an operation does not act correctly according to the documentation,
+or if you need a deterministic implementation of an operation that does not have
+one, please submit an issue:
+`<https://github.com/pytorch/pytorch/issues?q=label:%22topic:%20determinism%22>`_
 
-There are some PyTorch functions that use CUDA functions that can be a source
-of nondeterminism. One class of such CUDA functions are atomic operations,
-in particular :attr:`atomicAdd`, which can lead to the order of additions being
-nondetermnistic. Because floating-point addition is not perfectly associative
-for floating-point operands, :attr:`atomicAdd` with floating-point operands can
-introduce different floating-point rounding errors on each evaluation, which
-introduces a source of nondeterministic variance (aka noise) in the result.
+.. code-block::
 
-PyTorch functions that use :attr:`atomicAdd` in the forward kernels include
-:meth:`torch.Tensor.index_add_`, :meth:`torch.Tensor.scatter_add_`,
-:meth:`torch.bincount`.
+    torch.set_deterministic(True)
 
-A number of operations have backwards kernels that use :attr:`atomicAdd`,
-including :meth:`torch.nn.functional.embedding_bag`,
-:meth:`torch.nn.functional.ctc_loss`, :meth:`torch.nn.functional.interpolate`,
-and many forms of pooling, padding, and sampling.
+CUDA convolution
+----------------
+When running convolution operations CUDA devices, two options must be set::
 
-There is currently no simple way of avoiding nondeterminism in these functions.
-
-Additionally, the backward path for :meth:`repeat_interleave` operates
-nondeterministically on the CUDA backend because :meth:`repeat_interleave`
-is implemented using :meth:`index_select`, the backward path for
-which is implemented using :meth:`index_add_`, which is known to operate
-nondeterministically (in the forward direction) on the CUDA backend (see above).
+    torch.backends.cudnn.deterministic = True
```

Comment:
In the current introduction to this note it suggests three things will be detailed:

1) Setting PyTorch's psuedorandom generator
2) Configuring PyTorch to avoid nondeterministic algorithms
3) Setting the seeds of other libraries

And this section doesn't obviously fit into those three (two) topics. 

Setting `torch.backends.cudnn.benchmark = False` is controlling a source of nondeterminism, however, and the discussion of why it's necessary is interesting. 

Is a more accurate (if less descriptive) division simply:

1) Controlling sources of randomness
2) Avoiding nondeterministic algorithms

And then the first section would have the latitude to mention examples like setting the seeds in random generators and avoiding noisy benchmarks?

Reply:
I think these two divisions make sense. In that case, should I just remove the mention of `torch.backends.cudnn.deterministic`? I'm not sure if anyone would want to enable determinism for ONLY cuDNN anyway.

Reply:
Per offline discussion: let's find a home for the `torch.backends.cudnn.deterministic` discussion since it's useful information.

Reply:
Ok I've reorganized a few things. The document has this hierarchy now:

* Controlling sources of randomness
  * PyTorch random number generator
  * Random number generators in other libraries
  * CUDA convolution benchmarking
* Avoiding nondeterministic algorithms
  * CUDA convolution determinism
  * CUDA RNN and LSTM